mlflow/lightgbm.py

```diff
@@ -484,10 +489,23 @@ def log_feature_importance_plot(features, importance, importance_type):
 
             # If early_stopping_rounds is present, logging metrics at the best iteration
             # as extra metrics with the max step + 1.
-            early_stopping_index = all_arg_names.index("early_stopping_rounds")
-            early_stopping = (
-                num_pos_args >= early_stopping_index + 1 or "early_stopping_rounds" in kwargs
-            )
+            if Version(lightgbm.__version__) <= Version("3.3.1"):
+                # The parameter `early_stopping_rounds` in `lightgbm.train` is removed in this PR:
+                # https://github.com/microsoft/LightGBM/pull/4908
+                early_stopping_index = all_arg_names.index("early_stopping_rounds")
+                early_stopping = (
+                    num_pos_args >= early_stopping_index + 1 or "early_stopping_rounds" in kwargs
+                )
+            else:
+                early_stopping = False
+                if "callbacks" in kwargs and kwargs["callbacks"] is not None:
+                    for cb in kwargs["callbacks"]:
+                        if (
+                            hasattr(cb, "__qualname__")
+                            and cb.__qualname__ == "early_stopping.<locals>._callback"
+                        ):
+                            early_stopping = True
+                            break
             if early_stopping:
```

Comment:
https://github.com/microsoft/LightGBM/blob/ce486e5b45a6f5e67743e14765ed139ff8d532e5/python-package/lightgbm/engine.py#L226-L263
`best_iteration` is initiated as 0 and is only assigned with a positive int when `EarlyStoppingException` is raised (this means early stopping is activated, including using `early_stopping_rounds` param, `lightgbm.early_stopping` callback, or even user-defined early_stopping callbacks that raises `EarlyStoppingException`).
iff early stopping is not activated, `best_iteration` will be 0.